### ISSUE TITLE
Implement `--preview-dart-2` option for `pub run`

### ISSUE THREAD

**Issue by nex3** [dart-lang/pub#1807]
_From @aam on February 17, 2018 1:41_

Are there plans to implement `--preview-dart-2` option for `pub run`?

See https://github.com/flutter/flutter/pull/14728#discussion_r168903155 for context.

_Copied from original issue: dart-lang/sdk#32214_

**Comment by nex3** [comment]
_From @kevmoo on February 20, 2018 17:33_

Related to https://github.com/dart-lang/sdk/issues/32188

**Comment by nex3** [comment]
_From @kevmoo on February 20, 2018 17:34_

CC @a-siva

**Comment by nex3** [comment]
I'd love to add this, but we need support from the isolate library first. I've filed https://github.com/dart-lang/sdk/issues/32253 to track that.

**Comment by natebosch** [comment]
Could we have an API to ask pub for the entrypoint rather than have it run it itself?

That way we could do something like `dart --preview-dart-2 $(pub run --find_entrypoint some_package)`

**Comment by nex3** [comment]
We *could*, but that invocation wouldn't just work. The entrypoint is likely to be somewhere that automated package resolution won't work, so you'd have to run something more like `dart --preview-dart-2 --packages=.packages $(pub run --find-entrypoint some_package)`. I'd really like to provide a more streamlined option than that.

**Comment by natebosch** [comment]
I'd be super happy to have that option, even outside of the `--preview-dart-2` discussion. It would also give us an opportunity to run this code in our own isolate.

**Comment by nex3** [comment]
@natebosch Maybe we should discuss this on another issue? I'd be interested in hearing your use-cases.

**Comment by tvolkert** [comment]
Discussed with @a-siva, and it sounds like if pub were made to be Dart 2 compliant, then dart-lang/sdk#32253 wouldn't be needed, because pub could start in Dart 2, in which case all of its spawned isolates would start in Dart 2 as well.

If this is correct, I'm surprised that pub isn't _already_ Dart 2 compliant, with all the work that we've been asking of our ecosystem to migrate to Dart 2 compatibility.

/cc @leafpetersen

**Comment by nex3** [comment]
Dart 2 compliance for pub is blocked on Dart 2 compliance for dart2js and for the analyzer.

**Comment by kevmoo** [comment]
> Dart 2 compliance for pub is blocked on Dart 2 compliance for dart2js and for the analyzer.

After https://github.com/dart-lang/pub/issues/1857 – it'd just be analyzer, right?

**Comment by leafpetersen** [comment]
> If this is correct, I'm surprised that pub isn't already Dart 2 compliant, with all the work that we've been asking of our ecosystem to migrate to Dart 2 compatibility.

In fairness, most folks have been too busy helping the ecosystem migrate, but yes, it's more than a little embarrassing that we haven't migrated all of our own code yet.  :}

**Comment by stevenroose** [comment]
Just a related question: I have Dart v1 installed on my system and `dart` and `pub` are in `PATH` for regular commands. I also installed Dart v2 in `/opt/dart-v2`. Now, when I run `/opt/dart-v2/bin/pub run test` in a project, does it use the Dart v2 executable in `/opt/dart-v2/bin/dart`, or the default `dart` in `PATH`?

**Comment by natebosch** [comment]
The `pub` executable should use the Dart VM from it's own SDK. So if you run `/opt/dart-v2/bin/pub` you'll be using `/opt/dart-v2/bin/dart` vm.

**Comment by nex3** [comment]
To be clear, though, just running with a Dart VM from a 2.0.0-dev SDK doesn't give you full Dart 2 semantics. By default, you'll still be running in Dart 1 mode unless you pass the `--preview-dart-2` option to the VM, which currently doesn't work for pub.

**Comment by kevmoo** [comment]
I does work (a bit) w/ SDK @ HEAD. Enough to run pub get, pub run test, etc

**Comment by chalin** [comment]
Has this landed (in an externally released SDK)?

```console
> pub run -h
Run an executable from a package.

Usage: pub run <executable> [args...]
-h, --help            Print this usage information.
-c, --[no-]checked    Enable runtime type checks and assertions.

Run "pub help" to see global options.
```

That reminds me, `--[no-]checked` should be removed/replaced.

**Comment by kevmoo** [comment]
You can run `DART_VM_OPTIONS=--preview-dart-2 pub run...` with latest -dev

> That reminds me, --[no-]checked should be removed/replaced.

@chalin – please open an issue!

CC @grouma – might be good to hold v1 for this?

**Comment by chalin** [comment]
Done: #1922

**Comment by chalin** [comment]
Given that Dart 2 stable is out, this can be closed, right? /cc @kwalrath

**Comment by kevmoo** [comment]
Yup!
